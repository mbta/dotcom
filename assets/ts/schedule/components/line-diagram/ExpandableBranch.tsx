import { times } from "lodash";
import React, { ReactElement, useContext, useState } from "react";
import ExpandableBlock from "../../../components/ExpandableBlock";
import { routeToModeName } from "../../../helpers/css";
import { stopForId } from "../../../helpers/stop-tree";
import { Alert } from "../../../__v3api";
import { RouteStop, StopId, StopTree } from "../__schedule";
import {
  BRANCH_LINE_WIDTH,
  CIRC_DIAMETER,
  CIRC_RADIUS
} from "./graphics/graphic-helpers";
import { branchPosition, diagramWidth } from "./line-diagram-helpers";
import { StopRefContext } from "./LineDiagramWithStops";
import StopCard from "./StopCard";
import { LiveDataByStop } from "./__line-diagram";
import { alertsByStop } from "../../../models/alert";

interface Props {
  stopTree: StopTree;
  stopIds: StopId[];
  alerts: Alert[];
  handleStopClick: (stop: RouteStop) => void;
}

const BranchToggle = (
  stopTree: StopTree,
  stopIds: StopId[],
  isExpanded: boolean
): JSX.Element => {
  const position = branchPosition(stopTree, stopIds[0]);
  const width = diagramWidth(position) - 15;

  return (
    <div className="m-schedule-diagram__expander-header">
      <div style={{ width }}>
        {!isExpanded && (
          <svg
            className={`line-diagram-svg__toggle ${routeToModeName(
              stopForId(stopTree, stopIds[0]).route!
            )}`}
            width={width}
            height="100%"
          >
            <g
              transform={`translate(${(position - 1) * BRANCH_LINE_WIDTH -
                1}, 1)`}
            >
              <rect width={CIRC_DIAMETER * 2} height="42" rx={CIRC_DIAMETER} />
              {times(3, i => (
                <circle
                  key={i}
                  cx={CIRC_DIAMETER}
                  cy={10 * (i + 1) + 1}
                  r={CIRC_RADIUS}
                />
              ))}
            </g>
          </svg>
        )}
      </div>
      <button className="btn btn-link m-schedule-diagram__toggle" type="button">
        {stopIds.length} {stopIds.length > 1 ? "stops" : "stop"}
      </button>
    </div>
  );
};

const ExpandableBranch = ({
  stopTree,
  stopIds,
  alerts,
  handleStopClick,
}: Props): ReactElement<HTMLElement> => {
  const [isExpanded, setIsExpanded] = useState(false);
  const updateAllStopCoords = useContext(StopRefContext)[1];
  // reset all the coordinates.
  React.useEffect(() => {
    updateAllStopCoords();
  }, [isExpanded, updateAllStopCoords]);

  return (
    <div className="m-schedule-diagram__expander">
      <ExpandableBlock
        header={{
          iconSvgText: null,
          text: BranchToggle(stopTree, stopIds, isExpanded)
        }}
        initiallyExpanded={isExpanded}
        notifyExpanded={(blockIsExpanded: boolean) => {
          setIsExpanded(blockIsExpanded);
        }}
        preventScroll
        id={`${stopForId(stopTree, stopIds[0]).branch}`}
      >
        <>
          {stopIds.map(stopId => (
            <StopCard
              key={stopId}
              stopTree={stopTree}
              routeStopList={[]}
              stopId={stopId}
              alerts={alertsByStop(alerts, stopId)}
              onClick={handleStopClick}
            />
          ))}
        </>
      </ExpandableBlock>
    </div>
  );
};

export default ExpandableBranch;
