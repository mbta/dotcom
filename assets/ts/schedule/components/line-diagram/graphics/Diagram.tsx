import React, { ReactElement } from "react";
import { routeToModeName } from "../../../../helpers/css";
import {
  isBranchingNode,
  largestSliceSize,
  longestPath,
  longestPathLength,
  nextStopIds,
  stopForId,
  stopIds
} from "../../../../helpers/stop-tree";
import { isAGreenLineRoute } from "../../../../models/route";
import { Alert, DirectionId, Route } from "../../../../__v3api";
import {
  IndexedRouteStop,
  RouteStop,
  StopId,
  StopTree
} from "../../__schedule";
import { diagramWidth } from "../line-diagram-helpers";
import { BASE_LINE_WIDTH, DiagonalHatchPattern } from "./graphic-helpers";
import Stop from "./Stop";
import { Line, SimpleLine } from "./Line";
import Merges from "./Merges";

interface Props {
  stopTree: StopTree;
  route: Route;
  directionId: DirectionId;
  alerts: Alert[];
}

const routeName = (route: Route): string =>
  isAGreenLineRoute(route) ? "Green Line" : route.name;

const diagramDescription = (
  pathLength: number,
  route: Route,
  directionId: DirectionId
): string => {
  const text = `${pathLength} stops`;
  const {
    direction_destinations: destinations,
    direction_names: names
  } = route;
  if (destinations) {
    return `${text} going ${names[directionId]} to ${destinations[directionId]}`;
  }
  return text;
};

const hasBranchLines = (stopTree: StopTree): boolean =>
  largestSliceSize(stopTree) > 1;

const branchingDescription = (stopTree: StopTree): string => {
  const primaryPath = longestPath(stopTree);
  const branchingStops: StopId[] = primaryPath.filter(id =>
    isBranchingNode(stopTree, id)
  );
  const branchText = branchingStops.map(branchingStopId => {
    const routeStop: RouteStop = stopForId(stopTree, branchingStopId);
    const branchNames: string[] = nextStopIds(stopTree, branchingStopId)
      .filter(id => !primaryPath.includes(id))
      .map(id => stopForId(stopTree, id)?.branch || "");

    return `Use ${routeStop.name} to transfer to trains on ${branchNames.join(
      " or "
    )} branches.`;
  });
  return branchText.join("");
};

interface SimpleProps {
  routeStopList: IndexedRouteStop[];
  route: Route;
  directionId: DirectionId;
  alerts: Alert[];
}

const SimpleDiagram = ({
  routeStopList,
  route,
  directionId,
  alerts
}: SimpleProps): ReactElement<HTMLElement> => (
  <>
    <svg
      xmlns="http://www.w3.org/2000/svg"
      role="img"
      aria-labelledby="diagram-title diagram-desc"
      className={`line-diagram-svg ${routeToModeName(route)}`}
      width={`${diagramWidth(1)}px`}
      height="100%"
      style={{ left: BASE_LINE_WIDTH / 2, overflow: "visible" }}
    >
      <title id="diagram-title">Line diagram for {routeName(route)}</title>
      <desc id="diagram-desc">
        {diagramDescription(routeStopList.length, route, directionId)}
      </desc>
      <defs>{DiagonalHatchPattern()}</defs>
      {/* Draw lines between stops */
      routeStopList.map((routeStop, index) => {
        const nextStop = routeStopList[index + 1];
        if (!nextStop) return null;
        return (
          <SimpleLine
            key={`${routeStop.id}-${nextStop.id}`}
            fromId={routeStop.id}
            toId={nextStop.id}
            alerts={alerts}
          />
        );
      })}

      {/* Draw circles for each stop */
      routeStopList.map(routeStop => (
        <Stop key={`stop-${routeStop.id}`} stopId={routeStop.id} />
      ))}
    </svg>
  </>
);

const Diagram = ({
  stopTree,
  route,
  directionId,
  alerts
}: Props): ReactElement<HTMLElement> => (
  <>
    <svg
      xmlns="http://www.w3.org/2000/svg"
      role="img"
      aria-labelledby="diagram-title diagram-desc"
      className={`line-diagram-svg ${routeToModeName(route)}`}
      width={`${diagramWidth(1)}px`}
      height="100%"
      style={{ left: BASE_LINE_WIDTH / 2, overflow: "visible" }}
    >
      <title id="diagram-title">Line diagram for {routeName(route)}</title>
      <desc id="diagram-desc">
        {diagramDescription(longestPathLength(stopTree), route, directionId)}
        {hasBranchLines(stopTree) && branchingDescription(stopTree)}
      </desc>
      <defs>{DiagonalHatchPattern()}</defs>

      {/* If there are multiple branches, draw the lines and curves for those */
      hasBranchLines(stopTree) && (
        <Merges stopTree={stopTree} alerts={alerts} />
      )}

      {/* Draw lines between stops */
      stopIds(stopTree).map(stopId =>
        nextStopIds(stopTree, stopId).map(nextStopId => (
          <Line
            key={`${stopId}-${nextStopId}`}
            stopTree={stopTree}
            fromId={stopId}
            toId={nextStopId}
            alerts={alerts}
          />
        ))
      )}

      {/* Draw circles for each stop */
      stopIds(stopTree).map(stopId => (
        <Stop key={stopId} stopId={stopId} />
      ))}
    </svg>
  </>
);

export { Diagram, SimpleDiagram };
