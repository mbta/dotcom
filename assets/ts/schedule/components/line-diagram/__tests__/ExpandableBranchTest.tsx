import React from "react";
import * as redux from "react-redux";
import { mount, ReactWrapper } from "enzyme";
import { createStopTreeCoordStore } from "../graphics/useTreeStopPositions";
import { RouteStop, StopTree } from "../../__schedule";
import ExpandableBranch from "../ExpandableBranch";
import { RouteType } from "../../../../__v3api";

const stopTree: StopTree = {
  byId: {
    a1: {
      id: "a1",
      value: ({
        id: "a1",
        route: { id: "route-1", type: 3 },
        connections: [],
        stop_features: []
      } as unknown) as RouteStop
    },
    a2: {
      id: "a2",
      value: ({
        id: "a2",
        route: { id: "route-1", type: 3 },
        connections: [],
        stop_features: []
      } as unknown) as RouteStop
    },
    b1: {
      id: "b1",
      value: ({
        id: "b1",
        route: { id: "route-1", type: 3 },
        connections: [],
        stop_features: []
      } as unknown) as RouteStop
    },
    b2: {
      id: "b2",
      value: ({
        id: "b2",
        route: { id: "route-1", type: 3 },
        connections: [],
        stop_features: []
      } as unknown) as RouteStop
    },
    b3: {
      id: "b3",
      value: ({
        id: "b3",
        route: { id: "route-1", type: 3 },
        connections: [],
        stop_features: []
      } as unknown) as RouteStop
    },
    c1: {
      id: "c1",
      value: ({
        id: "c1",
        route: { id: "route-1", type: 3 },
        connections: [],
        stop_features: []
      } as unknown) as RouteStop
    },
    c2: {
      id: "c2",
      value: ({
        id: "c2",
        route: { id: "route-1", type: 3 },
        connections: [],
        stop_features: []
      } as unknown) as RouteStop
    },
    m1: {
      id: "m1",
      value: ({
        id: "m1",
        route: { id: "route-1", type: 3 },
        connections: [],
        stop_features: []
      } as unknown) as RouteStop
    },
    m2: {
      id: "m2",
      value: ({
        id: "m2",
        route: { id: "route-1", type: 3 },
        connections: [],
        stop_features: []
      } as unknown) as RouteStop
    },
    m3: {
      id: "m3",
      value: ({
        id: "m3",
        route: { id: "route-1", type: 3 },
        connections: [],
        stop_features: []
      } as unknown) as RouteStop
    },
    x1: {
      id: "x1",
      value: ({
        id: "x1",
        route: { id: "route-1", type: 3 },
        connections: [],
        stop_features: []
      } as unknown) as RouteStop
    },
    x2: {
      id: "x2",
      value: ({
        id: "x2",
        route: { id: "route-1", type: 3 },
        connections: [],
        stop_features: []
      } as unknown) as RouteStop
    },
    y1: {
      id: "y1",
      value: ({
        id: "y1",
        route: { id: "route-1", type: 3 },
        connections: [],
        stop_features: []
      } as unknown) as RouteStop
    }
  },
  edges: {
    a1: { next: ["a2"], previous: [] },
    a2: { next: ["m1"], previous: ["a1"] },
    b1: { next: ["b2"], previous: [] },
    b2: { next: ["b3"], previous: ["b1"] },
    b3: { next: ["m1"], previous: ["b2"] },
    c1: { next: ["c2"], previous: [] },
    c2: { next: ["m2"], previous: ["c1"] },
    m1: { next: ["m2"], previous: ["a2", "b3"] },
    m2: { next: ["m3"], previous: ["c2", "m1"] },
    m3: { next: ["x1", "y1"], previous: ["m2"] },
    x1: { next: ["x2"], previous: ["m3"] },
    x2: { next: [], previous: ["x1"] },
    y1: { next: [], previous: ["m3"] }
  },
  startingNodes: ["a1", "b1", "c1"]
};
const store = createStopTreeCoordStore(stopTree);

const handleStopClick = () => {};

describe("ExpandableBranch", () => {
  let wrapper: ReactWrapper;

  beforeEach(() => {
    wrapper = mount(
      <redux.Provider store={store}>
        <ExpandableBranch
          stopTree={stopTree}
          stopIds={["x1", "x2"]}
          alerts={[]}
          handleStopClick={handleStopClick}
        />
      </redux.Provider>
    );
  });

  afterEach(() => {
    wrapper.unmount();
  });

  it("renders and matches snapshot", () => {
    expect(wrapper.debug()).toMatchSnapshot();
  });

  it("expands a branch", () => {
    let moreStops = wrapper.find(
      ".c-expandable-block__panel .m-schedule-diagram__stop"
    );
    expect(moreStops.exists()).toBeFalsy();
    wrapper
      .find(
        ".m-schedule-diagram__expander .c-expandable-block__header[role='button']"
      )
      .first()
      .simulate("click");
    moreStops = wrapper.find(
      ".c-expandable-block__panel .m-schedule-diagram__stop"
    ); // update the reference so the test finds it
    expect(moreStops.exists()).toBeTruthy();
  });
});
