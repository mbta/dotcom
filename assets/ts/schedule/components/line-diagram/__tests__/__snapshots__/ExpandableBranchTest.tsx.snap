// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExpandableBranch renders and matches snapshot 1`] = `
"<Provider store={{...}}>
  <ExpandableBranch stopTree={{...}} stopIds={{...}} alerts={{...}} handleStopClick={[Function: handleStopClick]}>
    <div className=\\"m-schedule-diagram__expander\\">
      <ExpandableBlock header={{...}} initiallyExpanded={false} notifyExpanded={[Function: notifyExpanded]} preventScroll={true} id=\\"undefined\\">
        <h3 className=\\"c-expandable-block__header \\" tabIndex={0} id=\\"header-undefined\\" aria-expanded={false} aria-controls=\\"panel-undefined\\" role=\\"button\\" onClick={[Function: onClick]} onKeyPress={[Function: onKeyPress]}>
          <div className=\\"c-expandable-block__header-text\\">
            <div className=\\"m-schedule-diagram__expander-header\\">
              <div style={{...}}>
                <svg className=\\"line-diagram-svg__toggle bus\\" width={21} height=\\"100%\\">
                  <g transform=\\"translate(-1, 1)\\">
                    <rect width={16} height=\\"42\\" rx={8} />
                    <circle cx={8} cy={11} r={4} />
                    <circle cx={8} cy={21} r={4} />
                    <circle cx={8} cy={31} r={4} />
                  </g>
                </svg>
              </div>
              <button className=\\"btn btn-link m-schedule-diagram__toggle\\" type=\\"button\\">
                2
                 
                stops
              </button>
            </div>
            <span className=\\"fa c-expandable-block__header-caret\\" dangerouslySetInnerHTML={{...}} />
          </div>
        </h3>
        <noscript>
          <style />
          <h3 className=\\"c-expandable-block__header\\" />
          <div className=\\"c-expandable-block__panel\\" role=\\"region\\" />
        </noscript>
      </ExpandableBlock>
    </div>
  </ExpandableBranch>
</Provider>"
`;
